Apply refactorings

Shatur95 opened the pull request:
Closes #11.
I made this behavior by default because normally you never need to perform a checkout to the origin branches. Feel free to disagree with this.

Shatur95 commented on autoload/fzf_checkout.vim:
```diff
@@ -97,13 +97,17 @@ function! fzf_checkout#list(bang, type)
   if a:type ==# 'branch'
-    let l:subcommand = 'branch --all'
+    if g:fzf_checkout_hide_remote
+      let l:subcommand = 'branch --remote' " Show all remote branches to strip origin/ prefix later
```
Maybe use `const` instead of `let` and move outside from this function to allocate it only once?

Shatur95 commented on autoload/fzf_checkout.vim:
```diff
@@ -97,13 +97,17 @@ function! fzf_checkout#list(bang, type)
   " See valid atoms in
   " https://github.com/git/git/blob/076cbdcd739aeb33c1be87b73aebae5e43d7bcc5/ref-filter.c#L474
   let l:format =
-        \ '%(color:yellow bold)%(refname:short)  ' ..
+        \ '%(color:yellow bold)%(refname:' .. (g:fzf_checkout_hide_remote ? 'strip=3' : 'short') .. ')  ' ..
         \ '%(color:reset)%(color:green)%(subject) ' ..
         \ '%(color:reset)%(color:green dim italic)%(committerdate:relative) ' ..
         \ '%(color:reset)%(color:blue)-> %(objectname:short)'
```
Maybe use `const` instead of `let` and move outside from this function to allocate it only once?

stsewd replied:
That sounds good :+1:

Shatur95 commented:
Currently this does not work exactly as I want. It only shows remote branches without prefixes. I would like to show all local branches and non-duplicating remote branches without a prefix. I will try to figure it out.

Shatur95 commented:
The only way I found is call `git branch ---local` to get list of local branches and then call `git branch --remote` with strip=3 and merge without duplicates with local :(
I do not like this idea because it will not fit with this plugin. Is there another way to do it? If not, I will just create custom solution for my vimrc.

stsewd commented:
> The only way I found is call git branch ---local to get list of local branches and then call git branch --remote with strip=3 and merge without duplicates with local 

Well, that is the way I was thinking, but I also want to explore the idea of doing the filtering inside vim instead of concatenating all those bash commands p:

stsewd commented:
Sorry by taking so long to reply!

Shatur95 commented:
> Well, that is the way I was thinking, but I also want to explore the idea of doing the filtering inside vim instead of concatenating all those bash commands p:

What if get `git branch --all`, then on each item that contains `/` get tail (via `:t` like a path) and then remove duplicates?

Shatur95 commented:
I tried this approach. I also get rid of all other extra process executions (except getting list of branches/tag and current ref) to make all parsing inside Vim. Because of this I temporary I removed getting previous tag (I will put It back later).
It works, but I need to remove duplicates from the obtained list. Will try it tomorrow.

Shatur95 commented:
Now it works as I suggested :)

Shatur95 commented:
Previously branches was sorted by `--sort=refname:short` inside script. I moved it into an option and changed it `--sort=-committerdate` (with minus sign!). This will display recently used branches first.

Feel free to disagree with any my changed, waiting for your opinion :)

hanh090 commented on README.md:
````diff
@@ -124,5 +134,5 @@ let g:fzf_command_prefix = 'Fzf'
 
 ```vim
-let g:fzf_checkout_git_options = '--sort=committerdate'
+let g:fzf_checkout_git_options = '--sort=refname:short'
````
why do you change this config? it mismatch with text above

Shatur95 replied:
This is an example how it can be configured, not the default value. Previously `g:fzf_checkout_git_options` was empty and here `--sort=committerdate` was suggested. Currently `--sort=committerdate` is the default value in this PR, so I changed the example.

hanh090 replied:
then you should change the text above? because above text is `Sort branches/tags by committer date`

Shatur95 replied:
Oh, yes, thank you!

hanh090 commented on plugin/fzf_checkout.vim:
```diff
@@ -1,6 +1,6 @@
 " Global settings
 let g:fzf_checkout_git_bin = get(g:, 'fzf_checkout_git_bin', 'git')
-let g:fzf_checkout_git_options = get(g:, 'fzf_checkout_git_options', '')
+let g:fzf_checkout_git_options = get(g:, 'fzf_checkout_git_options', '--sort=-committerdate')
```
👍

Shatur95 commented:
Done!

stsewd commented:
>  What if get git branch --all, then on each item get tail (via fnamemodify with :t ) and then remove duplicates?

Is this the current approach of this PR? If so, there is a problem doing that, you can remove valid items, since git allows you to create branches with `/`, that's why I was trying to list locals and remotes by separate and then merging both lists.

Shatur95 commented:
> Is this the current approach of this PR? If so, there is a problem doing that, you can remove valid items, since git allows you to create branches with /, that's why I was trying to list locals and remotes by separate and then merging both lists.

Yes, did not thought about it...

stsewd commented:
I think with what you have done so far is going to be easier to merge both results :), but we should try to keep the original order, or just keep it simple for now and improve it later.

Shatur95 commented:
You are right. I will try to find another approach, maybe it somehow possible to get this via singe git command. If not, I will implement it as you suggested :)

Shatur95 commented:
> Is this the current approach of this PR? If so, there is a problem doing that, you can remove valid items, since git allows you to create branches with /, that's why I was trying to list locals and remotes by separate and then merging both lists.

What if just remove `remotes/any-remote-name/>`? So, branches with `\` in the name will be taken into account.

stsewd commented:
You can still have collisions, but I think it would be ok for now.

Shatur95 commented:
You are right, just checked remotes can have `/` in name too.
I will implement by two commands as you suggested.

Shatur95 commented:
Wait, I think that I a better way!
There is `for-each-ref`. If I pass `refs/heads` it prints just what I tried to do. It also accepts formatting. So, the following command do the trick (I copied formatting from the script to demonstrate how it will work):
```bash
git for-each-ref --format '%(color:yellow bold)%(refname:short)  %(color:reset)%(color:green)%(subject) %(color:reset)%(color:green dim italic)%(committerdate:relative) %(color:reset)%(color:blue)-> %(objectname:short)' refs/heads
```

So, what if just introduce `heads` type (we already have `tag` and `branch`)?

stsewd commented:
It should be a subcommand from `GCheckout`, something like `GCheckout heads` (or a name close to that) (assuming this only list branches). Also, looks like this only prints branches that are already being tracked, which is what `git branch` (without --all) already does, so if you want that behavior it should be `GCheckout locals`

Shatur95 commented:
> It should be a subcommand from GCheckout, something like GCheckout heads (or a name close to that) (assuming this only list branches).

Yes, that's exactly what I was thinking.

> Also, looks like this only prints branches that are already being tracked, which is what git branch (without --all) already does, so if you want that behavior it should be GCheckout locals.

You are right, I overlooked, this command doesn't show remote branches. So. no sense in it.

I think that calling two commands will not fit the plugin design. I looked at VSCode and it just start tracking branch remote branches automatically. Maybe we need to go in this way? Just add an option to check if branch is remote and start tracking it if yes and a do a simple checkout if not. For example, checkout to `origin/my-branch` will create local `my-branch`. This is how it works in VSCode.

stsewd commented:
> Just add an option to check if branch is remote and start tracking it if yes and a do a simple checkout if not

Yeah, that will work too. Currently, you can track a branch by pressing alt-enter, but doing it implicitly would be a great feature.

Shatur95 commented:
What if merge this branch as refactoring and add an option for automatic tracking in next PR?

At the moment, this branch works as the current master, but does not push the previous branch in list first. But this branch does not require `sed` and `print`. Also, what do you think if we leave this feature removed? Changed order may be not expected, especially if you use `--sort=refname:short`.

stsewd commented:
> What if merge this branch as refactoring and add an option for automatic tracking in next PR?

sure, that sounds good

> but does not push the previous branch in list first. What do you think if we leave the feature removed? Changed order may be not expected, especially if you use --sort=refname:short.

That shouldn't be removed, we can add a setting for it. Also, we shouldn't change the defaults in this PR (`fzf_checkout_git_options`)

Shatur95 commented:
> That shouldn't be removed, we can add a setting for it. 

Good idea, I will add an option and make it enabled by default.

> Also, we shouldn't change the defaults in this PR (fzf_checkout_git_options)

If I understood correctly, `--sort=refname:short` was hardcoded [inside the script](https://github.com/stsewd/fzf-checkout.vim/blob/a352086722af18073bc9a1591473d398cae9bc72/autoload/fzf_checkout.vim#L115). I just moved it into `fzf_checkout_git_options`. Is it okay?

stsewd commented:
> If I understood correctly, --sort=refname:short was hardcoded inside the script. I just moved it into fzf_checkout_git_options. Is it okay?

`--sort` can be overridden, so isn't a problem having it there. And gives the user the ability to override other stuff without repeating the sort option.

Shatur95 commented:
Oh, this was expected? Yes, I knew that this option can be overwritten, I just thought that is better when it is clearly visible in options. Should I put it back too?

stsewd commented:
yes please

Shatur95 commented:
Done! Could you check it?

Shatur95 commented:
I also switched to `match` and `remove` which is more efficient than `map` for finding only first occurrence.

stsewd commented:
The only thing I have experienced so far is that this version is slower than before, is really noticeable when you are in a repo with a lot of branches. I'm fine with it, but we definitely need to see if we can optimize this

Shatur95 commented:
This is strange, I was sure it would increase performance. Perhaps this is due to the fact that I use regex for the color too (for more reliability). I can use a hardcoded value, but could you please try with the last change first?

stsewd commented:
Much faster now! Great work @Shatur95!
